src/block_cache/block_chunkcache.c

```diff
@@ -0,0 +1,561 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *  All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#ifdef HAVE_DIAGNOSTIC
+#define WT_BLOCK_OVERLAPS_CHUNK(chunk_off, block_off, chunk_size, block_size) \
+    (block_off < chunk_off + (wt_off_t)chunk_size) && (chunk_off < block_off + (wt_off_t)block_size)
+#endif
+
+/* This rounds down to the chunk boundary. */
+#define WT_CHUNK_OFFSET(chunkcache, offset) \
+    (wt_off_t)(((size_t)offset / (chunkcache)->chunk_size) * (chunkcache)->chunk_size)
+
+#define WT_BUCKET_CHUNKS(chunkcache, bucket_id) &(chunkcache->hashtable[bucket_id].colliding_chunks)
+#define WT_BUCKET_LOCK(chunkcache, bucket_id) &(chunkcache->hashtable[bucket_id].bucket_lock)
+
+/*
+ * __chunkcache_alloc --
+ *     Allocate memory for the chunk in the cache.
+ */
+static int
+__chunkcache_alloc(WT_SESSION_IMPL *session, WT_CHUNKCACHE_CHUNK *chunk)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_DECL_RET;
+
+    chunkcache = &S2C(session)->chunkcache;
+
+    if (chunkcache->type == WT_CHUNKCACHE_IN_VOLATILE_MEMORY)
+        ret = __wt_malloc(session, chunk->chunk_size, &chunk->chunk_memory);
+    else {
+#ifdef ENABLE_MEMKIND
+        chunk->chunk_memory = memkind_malloc(chunkcache->memkind, chunk->chunk_size);
+        if (chunk->chunk_memory == NULL)
+            ret = ENOMEM;
+#else
+        WT_RET_MSG(session, EINVAL,
+          "Chunk cache requires libmemkind, unless it is configured to be in DRAM");
+#endif
+    }
+    if (ret == 0) {
+        __wt_atomic_add64(&chunkcache->bytes_used, chunk->chunk_size);
+        WT_STAT_CONN_INCR(session, chunk_cache_chunks);
+        WT_STAT_CONN_INCRV(session, chunk_cache_bytes, chunk->chunk_size);
+    }
+    return (ret);
+}
+
+/*
+ * __chunkcache_admit_size --
+ *     Decide if we can admit the chunk given the limit on cache capacity and return the size of the
+ *     chunk to be admitted.
+ */
+static size_t
+__chunkcache_admit_size(WT_SESSION_IMPL *session)
+{
+    WT_CHUNKCACHE *chunkcache;
+
+    chunkcache = &S2C(session)->chunkcache;
+
+    if ((chunkcache->bytes_used + chunkcache->chunk_size) < chunkcache->capacity)
+        return (chunkcache->chunk_size);
+
+    WT_STAT_CONN_INCR(session, chunk_cache_exceeded_capacity);
+    __wt_verbose(session, WT_VERB_CHUNKCACHE, "exceeded chunkcache capacity of %" PRIu64 " bytes",
+      chunkcache->capacity);
+    return (0);
+}
+
+/*
+ * __chunkcache_alloc_chunk --
+ *     Allocate the chunk and its metadata for a block at a given offset.
+ */
+static int
+__chunkcache_alloc_chunk(WT_SESSION_IMPL *session, wt_off_t offset, WT_BLOCK *block,
+  WT_CHUNKCACHE_HASHID *hash_id, WT_CHUNKCACHE_CHUNK **newchunk)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_DECL_RET;
+    size_t chunk_size;
+    uint64_t hash;
+
+    *newchunk = NULL;
+    chunkcache = &S2C(session)->chunkcache;
+
+    WT_ASSERT(session, offset > 0);
```

Comment:
Are we holding a lock when we call this? If so, is there an assert we can/should add here to confirm this?

Reply:
I added a comment. Didn't see anything equivalent to lock_do_I_hold API, so I used trylock to make sure we are holding the lock. Didn't see this use of trylock anywhere in WiredTiger, but happy to break the tradition, if you find thing useful.